package/kedro_viz/data_access/repositories/modular_pipelines.py

```diff
@@ -83,163 +128,201 @@ def get_or_create_modular_pipeline(
-            )
+        hashed_inputs = set()
 
-        is_internal_input = modular_pipeline_id in input_node.modular_pipelines
-        if is_internal_input:
-            self.tree[modular_pipeline_id].internal_inputs.add(input_node.id)
-        else:
-            self.tree[modular_pipeline_id].external_inputs.add(input_node.id)
+        for _input in inputs:
+            hashed_input = _hash_input_output(_input)
+            hashed_inputs.add(hashed_input)
+            if is_dataset_param(_input):
+                self.parameters.add(hashed_input)
 
-    def add_output(self, modular_pipeline_id: str, output_node: GraphNode):
-        """Add an output to a modular pipeline based on whether it's an internal or external output.
-        The output has knowledge of which modular pipelines it belongs to based on its namespace.
-        The information can be accessed with through `output_node.modular_pipelines`.
+        self.tree[modular_pipeline_id].inputs = hashed_inputs
+
+    def _add_outputs(self, modular_pipeline_id: str, outputs: Set[str]) -> None:
+        """
+        Add output datasets from the modular pipeline.
 
         Args:
-            modular_pipeline_id: ID of the modular pipeline to add the output to.
-            output_node: The output node to add.
-        Raises:
-            ValueError: when attempt to add a non-data, non-parameter node as output
-                to the modular pipeline.
-        Example:
-            >>> modular_pipelines = ModularPipelinesRepository()
-            >>> data_science_pipeline = modular_pipelines.get_or_create_modular_pipeline(
-            ...     "data_science"
-            ... )
-            >>> model_output_node = GraphNode.create_data_node(
-            ...     "data_science.model_output", layer=None, tags=set(), dataset=None
-            ... )
-            >>> modular_pipelines.add_output("data_science", model_output_node)
-            >>> assert data_science_pipeline.outputs == {model_output_node.id}
+            modular_pipeline_id (str): The ID of the modular pipeline to add outputs to.
+            outputs (Set[str]): The output datasets to add.
         """
-        if not isinstance(output_node, (DataNode, TranscodedDataNode, ParametersNode)):
-            raise ValueError(
-                f"Attempt to add a non-data node as input to modular pipeline {modular_pipeline_id}"
-            )
+        hashed_outputs = {_hash_input_output(output) for output in outputs}
+        self.tree[modular_pipeline_id].outputs = hashed_outputs
 
-        is_internal_output = modular_pipeline_id in output_node.modular_pipelines
-        if is_internal_output:
-            self.tree[modular_pipeline_id].internal_outputs.add(output_node.id)
-        else:
-            self.tree[modular_pipeline_id].external_outputs.add(output_node.id)
+    def _add_children(self, modular_pipeline_id: str, kedro_nodes: List[Node]):

```

Comment:
This whole function is overly complicated, very wide open if condition and too much nestedness. Could you try to refactor it into a more concise and easier to follow one?